ISSUE-42: Auto create Security Groups

ossarga opened the pull request:
Closes #42 
Closes #36

rustyrazorblade commented on src/main/kotlin/com/thelastpickle/tlpcluster/terraform/Configuration.kt:
```diff
@@ -92,9 +93,43 @@ class Configuration(val tags: MutableMap<String, String> = mutableMapOf(),
+            .newSecurityGroupResource("${ticket}_TlpClusterSG","tlp-cluster ${ticket} security group", tags)
+            .withRule(0, 65535, "tcp", listOf("0.0.0.0/0"), "All traffic", SecurityGroupRule.Direction.Outbound)
+            .withRule(22, 22, "tcp", listOf("0.0.0.0/0"), "SSH", SecurityGroupRule.Direction.Inbound)
+            .withRule(7000, 7001, "tcp", listOf("172.31.0.0/16"), "Intra node", SecurityGroupRule.Direction.Inbound)
```
I don't think we should be using hard coded IPs here.  Ideally we'd be able to provide the security group, not an IP address.  This can be done through the UI and I think I've done it before with the API.

rustyrazorblade replied:
I believe what we need here is "self" for all the open ports within the SG:

https://www.terraform.io/docs/providers/aws/r/security_group_rule.html#self

ossarga replied:
Ok. I'll give it a try and rework the code once I have it working in a handcrafted terraform JSON file. I assume `self` will allow only the IPs within the VPC? (which is what we would want). If so, I guess we just give a port range and set the `self` property to `true`?

rustyrazorblade replied:
I was looking through this PR again and realized it's difficult to understand which rules are inbound and outbound.  We might want convenience methods `withInboundRule()` and `withOutboundRule()` to make it more obvious what's going on.  Just a thought.

ossarga replied:
Yeah good point.

rustyrazorblade commented:
I did it already in this branch, give it a look:
https://github.com/thelastpickle/tlp-cluster/compare/jon/42_sg_suggestions?expand=1

On Sun, Mar 24, 2019 at 3:08 PM Anthony Grasso <notifications@github.com>
wrote:

> *@ossarga* commented on this pull request.
> ------------------------------
>
> In src/main/kotlin/com/thelastpickle/tlpcluster/terraform/Configuration.kt
> <https://github.com/thelastpickle/tlp-cluster/pull/62#discussion_r268455582>
> :
>
> > @@ -92,9 +93,43 @@ class Configuration(val tags: MutableMap<String, String> = mutableMapOf(),
>          setVariable("region", region)
>          setVariable("zones", Variable(listOf("us-west-2a", "us-west-2b", "us-west-2c"), "list"))
>
> -        setResource("cassandra", cassandraAMI, cassandraInstanceType, numCassandraInstances, setTagName(tags, ServerType.Cassandra))
> -        setResource("stress", stressAMI, stressInstanceType, numStressInstances, setTagName(tags, ServerType.Stress))
> -        setResource("monitoring", monitoringAMI, monitoringInstanceType, if (monitoring) 1 else 0, setTagName(tags, ServerType.Monitoring))
> +        val instanceSg = SecurityGroupResource.Builder()
> +            .newSecurityGroupResource("${ticket}_TlpClusterSG","tlp-cluster ${ticket} security group", tags)
> +            .withRule(0, 65535, "tcp", listOf("0.0.0.0/0"), "All traffic", SecurityGroupRule.Direction.Outbound)
> +            .withRule(22, 22, "tcp", listOf("0.0.0.0/0"), "SSH", SecurityGroupRule.Direction.Inbound)
> +            .withRule(7000, 7001, "tcp", listOf("172.31.0.0/16"), "Intra node", SecurityGroupRule.Direction.Inbound)
>
> Ok. I'll give it a try and rework the code once I have it working in a
> handcrafted terraform JSON file. I assume self will allow only the IPs
> within the VPC? (which is what we would want). If so, I guess we just give
> a port range and set the self property to true?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/thelastpickle/tlp-cluster/pull/62#discussion_r268455582>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAUn0IsFxhEZxFf3PWPQrgww3VbKFPRoks5vZ_dUgaJpZM4cAxF_>
> .
>

ossarga commented:
@rustyrazorblade I have pushed up a new change that incorporates all feedback.
* Changed `withRule` function names so that it is clear if you are adding an inbound or outbound rule.
* Added `self` to SecurityGroupRule class and a `with` function to add a rule that uses it.
* Added code to get your external IP address and apply it to the SSH and monitoring inbound rules.

rustyrazorblade commented:
I think your last commit removing the security group introduced a bug, now it doesn't build.

ossarga commented:
Fixed!
